File: backend/app/routes/carpetas.py
from flask import request, jsonify, Blueprint
from ..models import User, Carpeta, Indiciado
from ..extensions import db
from flask_jwt_extended import jwt_required, get_jwt_identity

carpetas_bp = Blueprint('carpetas', __name__)

# --- CREAR CARPETA (Sin cambios) ---
@carpetas_bp.route('', methods=['POST'])
@jwt_required()
def crear_carpeta():
    current_username = get_jwt_identity()
    user = User.query.filter_by(username=current_username).first_or_404()
    
    data = request.get_json()
    if not data:
        return jsonify({"msg": "Cuerpo de la solicitud JSON vacío o inválido"}), 400

    nombre_carpeta = data.get('nombre')
    parent_id = data.get('parent_id')

    if not nombre_carpeta or not isinstance(nombre_carpeta, str) or nombre_carpeta.strip() == "":
        return jsonify({"msg": "El campo 'nombre' es requerido y debe ser un texto no vacío"}), 400
    
    nombre_carpeta = nombre_carpeta.strip()

    if parent_id:
        parent = Carpeta.query.filter_by(id=parent_id, user_id=user.id).first()
        if not parent:
            return jsonify({"msg": "El sector padre no existe o no pertenece al usuario"}), 404
        
        existe = Carpeta.query.filter_by(user_id=user.id, parent_id=parent_id, nombre=nombre_carpeta).first()
        if existe:
            return jsonify({"msg": f"El sub-sector '{nombre_carpeta}' ya existe en '{parent.nombre}'"}), 409

        nueva_carpeta = Carpeta(nombre=nombre_carpeta, user_id=user.id, parent_id=parent_id)

    else:
        existe = Carpeta.query.filter_by(user_id=user.id, parent_id=None, nombre=nombre_carpeta).first()
        if existe:
            return jsonify({"msg": f"El sector '{nombre_carpeta}' ya existe"}), 409
            
        nueva_carpeta = Carpeta(nombre=nombre_carpeta, owner=user)

    db.session.add(nueva_carpeta)
    db.session.commit()
    
    return jsonify({
        "msg": "Carpeta creada exitosamente",
        "carpeta": nueva_carpeta.to_dict()
    }), 201

@carpetas_bp.route('', methods=['GET'])
@jwt_required()
def obtener_carpetas_usuario():
    current_username = get_jwt_identity()
    user = User.query.filter_by(username=current_username).first_or_404()
    carpetas = Carpeta.query.filter_by(user_id=user.id, parent_id=None).order_by(Carpeta.nombre).all()
    return jsonify([c.to_dict() for c in carpetas]), 200

@carpetas_bp.route('/<int:id_carpeta>', methods=['GET'])
@jwt_required()
def obtener_carpeta_con_indiciados(id_carpeta):
    current_username = get_jwt_identity()
    user = User.query.filter_by(username=current_username).first_or_404()
    carpeta = Carpeta.query.filter_by(id=id_carpeta, user_id=user.id).first()
    
    if not carpeta:
        return jsonify({"msg": "Carpeta no encontrada o no pertenece al usuario"}), 404

    return jsonify(carpeta.to_dict(include_indiciados=True))

@carpetas_bp.route('/<int:id_carpeta>', methods=['PUT'])
@jwt_required()
def actualizar_carpeta(id_carpeta):
    current_username = get_jwt_identity()
    user = User.query.filter_by(username=current_username).first_or_404()
    
    carpeta = Carpeta.query.filter_by(id=id_carpeta, user_id=user.id).first()
    if not carpeta:
        return jsonify({"msg": "Carpeta no encontrada o no pertenece al usuario"}), 404

    data = request.get_json()
    nuevo_nombre = data.get('nombre')

    if not nuevo_nombre or not isinstance(nuevo_nombre, str) or nuevo_nombre.strip() == "":
        return jsonify({"msg": "El campo 'nombre' es requerido y debe ser un texto no vacío"}), 400

    nuevo_nombre = nuevo_nombre.strip()

    query_conflicto = Carpeta.query.filter(
        Carpeta.user_id == user.id,
        Carpeta.parent_id == carpeta.parent_id,
        Carpeta.nombre == nuevo_nombre,
        Carpeta.id != id_carpeta
    )
    conflicto = query_conflicto.first()

    if conflicto:
        return jsonify({"msg": f"Ya existe una carpeta con el nombre '{nuevo_nombre}' en este nivel."}), 409

    carpeta.nombre = nuevo_nombre
    db.session.commit()

    return jsonify({
        "msg": "Carpeta actualizada exitosamente",
        "carpeta": carpeta.to_dict()
    }), 200

@carpetas_bp.route('/<int:id_carpeta>', methods=['DELETE'])
@jwt_required()
def borrar_carpeta(id_carpeta):
    current_username = get_jwt_identity()
    user = User.query.filter_by(username=current_username).first_or_404()
    carpeta = Carpeta.query.filter_by(id=id_carpeta, user_id=user.id).first()
    if not carpeta:
        return jsonify({"msg": "Carpeta no encontrada o no pertenece al usuario"}), 404
    db.session.delete(carpeta)
    db.session.commit()
    return jsonify({"msg": "Carpeta eliminada exitosamente"}), 200

File: mern-backend/routes/carpetas.js
const express = require('express');
const { body, param, validationResult } = require('express-validator');
const Carpeta = require('../models/Carpeta');
const User = require('../models/User');
const Indiciado = require('../models/Indiciado');
const authMiddleware = require('../middleware/auth');

const router = express.Router();

// Crear carpeta
router.post('', authMiddleware, [
  body('nombre')
    .trim()
    .notEmpty()
    .withMessage('El campo "nombre" es requerido y debe ser un texto no vacío')
    .isLength({ max: 100 })
    .withMessage('El nombre no puede exceder 100 caracteres'),
  body('parent_id')
    .optional()
    .isMongoId()
    .withMessage('parent_id debe ser un ID válido')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        msg: 'Datos de validación incorrectos',
        details: errors.array()
      });
    }

    const { nombre, parent_id } = req.body;

    // Si hay parent_id, verificar que existe y pertenece al usuario
    if (parent_id) {
      const parent = await Carpeta.findOne({ 
        _id: parent_id, 
        ownerId: req.user._id 
      });
      
      if (!parent) {
        return res.status(404).json({
          msg: 'El sector padre no existe o no pertenece al usuario'
        });
      }

      // Verificar que no existe una carpeta con el mismo nombre en el mismo nivel
      const exists = await Carpeta.findOne({
        ownerId: req.user._id,
        parentId: parent_id,
        nombre
      });

      if (exists) {
        return res.status(409).json({
          msg: `El sub-sector '${nombre}' ya existe en '${parent.nombre}'`
        });
      }
    } else {
      // Verificar que no existe una carpeta raíz con el mismo nombre
      const exists = await Carpeta.findOne({
        ownerId: req.user._id,
        parentId: null,
        nombre
      });

      if (exists) {
        return res.status(409).json({
          msg: `El sector '${nombre}' ya existe`
        });
      }
    }

    // Crear nueva carpeta
    const nuevaCarpeta = new Carpeta({
      nombre,
      ownerId: req.user._id,
      parentId: parent_id || null
    });

    await nuevaCarpeta.save();
    const carpetaDict = await nuevaCarpeta.toDict();

    res.status(201).json({
      msg: 'Carpeta creada exitosamente',
      carpeta: carpetaDict
    });
  } catch (error) {
    console.error('Error creating carpeta:', error);
    res.status(500).json({
      error: 'Server error creating carpeta',
      message: error.message
    });
  }
});

// Obtener carpetas del usuario (solo carpetas raíz)
router.get('', authMiddleware, async (req, res) => {
  try {
    const carpetas = await Carpeta.find({
      ownerId: req.user._id,
      parentId: null
    }).sort({ nombre: 1 });

    const carpetasDict = [];
    for (const carpeta of carpetas) {
      const carpetaDict = await carpeta.toDict();
      carpetasDict.push(carpetaDict);
    }

    res.json(carpetasDict);
  } catch (error) {
    console.error('Error fetching carpetas:', error);
    res.status(500).json({
      error: 'Server error fetching carpetas',
      message: error.message
    });
  }
});

// Obtener carpeta específica con indiciados
router.get('/:id', authMiddleware, [
  param('id').isMongoId().withMessage('ID debe ser un ObjectId válido')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        msg: 'ID inválido',
        details: errors.array()
      });
    }

    const carpeta = await Carpeta.findOne({
      _id: req.params.id,
      ownerId: req.user._id
    });

    if (!carpeta) {
      return res.status(404).json({
        msg: 'Carpeta no encontrada o no pertenece al usuario'
      });
    }

    const carpetaDict = await carpeta.toDict(true);
    res.json(carpetaDict);
  } catch (error) {
    console.error('Error fetching carpeta:', error);
    res.status(500).json({
      error: 'Server error fetching carpeta',
      message: error.message
    });
  }
});

// Actualizar carpeta
router.put('/:id', authMiddleware, [
  param('id').isMongoId().withMessage('ID debe ser un ObjectId válido'),
  body('nombre')
    .trim()
    .notEmpty()
    .withMessage('El campo "nombre" es requerido y debe ser un texto no vacío')
    .isLength({ max: 100 })
    .withMessage('El nombre no puede exceder 100 caracteres')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        msg: 'Datos de validación incorrectos',
        details: errors.array()
      });
    }

    const { nombre } = req.body;
    const carpetaId = req.params.id;

    const carpeta = await Carpeta.findOne({
      _id: carpetaId,
      ownerId: req.user._id
    });

    if (!carpeta) {
      return res.status(404).json({
        msg: 'Carpeta no encontrada o no pertenece al usuario'
      });
    }

    // Verificar que no existe otra carpeta con el mismo nombre en el mismo nivel
    const conflicto = await Carpeta.findOne({
      ownerId: req.user._id,
      parentId: carpeta.parentId,
      nombre,
      _id: { $ne: carpetaId }
    });

    if (conflicto) {
      return res.status(409).json({
        msg: `Ya existe una carpeta con el nombre '${nombre}' en este nivel.`
      });
    }

    // Actualizar carpeta
    carpeta.nombre = nombre;
    await carpeta.save();

    const carpetaDict = await carpeta.toDict();

    res.json({
      msg: 'Carpeta actualizada exitosamente',
      carpeta: carpetaDict
    });
  } catch (error) {
    console.error('Error updating carpeta:', error);
    res.status(500).json({
      error: 'Server error updating carpeta',
      message: error.message
    });
  }
});

// Eliminar carpeta
router.delete('/:id', authMiddleware, [
  param('id').isMongoId().withMessage('ID debe ser un ObjectId válido')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        msg: 'ID inválido',
        details: errors.array()
      });
    }

    const carpeta = await Carpeta.findOne({
      _id: req.params.id,
      ownerId: req.user._id
    });

    if (!carpeta) {
      return res.status(404).json({
        msg: 'Carpeta no encontrada o no pertenece al usuario'
      });
    }

    // Las carpetas se eliminarán en cascada gracias a los middleware de Mongoose
    await Carpeta.findByIdAndDelete(req.params.id);

    res.json({
      msg: 'Carpeta eliminada exitosamente'
    });
  } catch (error) {
    console.error('Error deleting carpeta:', error);
    res.status(500).json({
      error: 'Server error deleting carpeta',
      message: error.message
    });
  }
});

module.exports = router;
